### PR TITLE
more instances for (Digest a)

### PR DESCRIPTION
**Pull request by paolino**
implementing Merkle root I needed 
`hash (x <> y)`

### CODE REVIEW

**Comment by paolino** [comment]
Read instance for `Digest a` in Base16 as Show instance.
read check both encoding and length.
I had to move `digestFromByteString` to Hash.Types, but it's still exported from Hash

**Comment by vincenthz** [comment]
@paolino: this can't be right; this allow creation of, `Digest a` which could be of any size, which break the assumption that it represent a digest for the algorithm `a`

Why not simply use `hashFinalize $ foldl' hashUpdate hashInit [x,y]` ?

**Comment by paolino** [comment]
You are right. If you agree that Digest a is a Monoid under HashAlgorithm a, I've commited the correct (I hope) instance (orphan) in Crypto.Hash

**Comment by vincenthz** [comment]
I don't think `Monoid` make sense for `Digest`. I think it breaks pretty much all laws for monoid. For starter:

```
mappend x mempty /= x
```

**Comment by paolino** [comment]
I see, there is no mempty for Digest a. I hope Read is not so wrong.
Thanks

**Comment by vincenthz** [comment]
On the other hand, `Read` is clearly useful, but I don't think it's correct. IIRC, it's eating everything until the end of the `String`, then calling digestFromByteString.

It should compute how much data is needed (known length of a `Digest`), then eat it, then unbase16 it to a digest.

**Comment by paolino** [comment]
I've fixed it.
I don't know if the (*2) is intended.
I think I'm hiding a possible implementation error in a Nothing, it should be "The impossible happened...." on the last bind, if digestFromByteString can only fail for short lengths

File: cryptonite.cabal [hunk 1]
```diff
@@ -192,7 +192,7 @@ Library
                    , bytestring
                    , memory >= 0.8
                    , ghc-prim
-  ghc-options:       -Wall -fwarn-tabs -optc-O3 -fno-warn-unused-imports
+  ghc-options:       -Wall -fwarn-tabs -optc-O3 -fno-warn-unused-imports -fobject-code
```
Comment (vincenthz) [inline cryptonite.cabal:195]: this shouldn't be here

**Comment by ixmatus** [comment]
I wanted to drop in and say I'd like to see the Read instance since I just implemented something similar by-hand.

**Comment by centromere** [comment]
@paolino Are you still interested in pursuing this pull request?

**Comment by osa1** [comment]
:+1: we also need a `Read` instance.

**Comment by ocheron** [comment]
Closing: `Monoid` part is not needed and `Read` part is replaced by #219.

**Pull request closed without merging**